File: Examples/Day04.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AdventOfCode.Examples
{
    public static class Day04
    {
        private static Dictionary<string, string> ToPassport(string value)
        {
            return value
                .Split(" ", StringSplitOptions.RemoveEmptyEntries)
                .Select(prop => prop.Split(":", StringSplitOptions.RemoveEmptyEntries))
                .ToDictionary(key => key[0], value => value[1]);
        }

        private static IEnumerable<Dictionary<string, string>> GetPassports()
        {
            return File.ReadAllText("Data/day04.txt")
                .Split(Environment.NewLine + Environment.NewLine)
                .Select(str => str.Replace(Environment.NewLine, " "))
                .Select(ToPassport)
                .ToList();
        }

        public static void Example01()
        {
            var passports = GetPassports();
            var requiredFields = new[] { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
            var validPassports = passports
                .Where(p => p.Keys.ContainsAll(requiredFields))
                .Count();

            Console.WriteLine(validPassports);
        }

        private static bool IsHeight(string value)
        {
            if (value.EndsWith("cm")) return value.WithoutPrefix("cm").Between(150, 193);
            if (value.EndsWith("in")) return value.WithoutPrefix("in").Between(59, 76);
            return false;
        }

        private static bool ValidateRule(Dictionary<string, string> values, Dictionary<string, Func<string, bool>> rules)
        {
            foreach (var value in values)
            {
                var rule = rules.GetValueOrDefault(value.Key, null);
                if (rule != null && !rule(value.Value))
                {
                    return false;
                }
            }

            return true;
        }

        public static void Example02()
        {
            var requiredFields = new[] { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
            var rules = new Dictionary<string, Func<string, bool>>
            {
                { "byr", value => value.Between(1920, 2002) },
                { "iyr", value => value.Between(2010, 2020) },
                { "eyr", value => value.Between(2020, 2030) },
                { "hgt", value => IsHeight(value) },
                { "hcl", value => value.Matches("#[a-f0-9]{6}") },
                { "ecl", value => new[] { "amb", "blu", "brn", "gry", "grn", "hzl", "oth" }.Contains(value) },
                { "pid", value => value.Length == 9 && value.Matches("[0-9]{9}") }
            };

            var validPassports = GetPassports()
                .Where(p => p.Keys.ContainsAll(requiredFields))
                .Where(values => ValidateRule(values, rules))
                .ToList();

            Console.WriteLine(validPassports.Count);
        }
    }
}

File: Swift/day04/day04.swift
import Foundation

extension Array where Element == String {
  func toMap(separatedBy str: String) -> [String: String] {
      return self.reduce(into: [:]) { result, next in
          let pair = next.components(separatedBy: str)
          result[pair[0]] = pair[1]
      }
  }
}

extension String {
    func matches(_ regex: String) -> Bool {
        return self.range(of: regex, options: .regularExpression, range: nil, locale: nil) != nil
    }
}

typealias Passport = [String: String]

func day04() {
  let requiredFields: Set = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
  let passports: [Passport] = readContent(from: "day04/input", ofType: "txt")
    .components(separatedBy: "\n\n")
    .map { $0
      .replacingOccurrences(of: "\n", with: " ")
      .components(separatedBy: " ")
      .toMap(separatedBy: ":")
    }

  // Example 01
  let validPassports = passports
    .filter { requiredFields.isSubset(of: $0.keys) }
    .count

  print("Valid passports 01: \(validPassports)")

  // Example 02
  let validPassports2 = passports
    .filter { requiredFields.isSubset(of: $0.keys) }
    .map { passport in
      passport.map { pair -> Bool in
        switch pair.key {
        case "byr": return (1920...2002).contains(Int(pair.value) ?? -1)
        case "iyr": return (2010...2020).contains(Int(pair.value) ?? -1)
        case "eyr": return (2020...2030).contains(Int(pair.value) ?? -1)
        case "hcl": return pair.value.matches("#[a-f0-9]{6}")
        case "pid": return pair.value.length == 9 && pair.value.matches("[0-9]{9}")
        case "ecl": return Set(["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]).contains(pair.value)
        case "hgt":
          switch pair.value.suffix(2) {
          case "cm": return (150...193).contains(Int(pair.value.dropLast(2)) ?? -1)
          case "in": return (59...76).contains(Int(pair.value.dropLast(2)) ?? -1)
          default: return false
          }
        default: return true
        }
      }
    }
    .filter { !$0.contains(false) }
    .count

  print("Valid passports 02: \(validPassports2)")
}